packages/engine/lib/execution/src/runner/javascript/mod.rs

```diff
@@ -885,13 +885,16 @@ impl<'s> ThreadLocalRunner<'s> {
         &self,
         data_ptr: NonNull<T>,
         data_len: usize,
-        _data_capacity: usize, // for future use to create a `Buffer::from_raw_parts`
         target_len: usize,
     ) -> Buffer {
-        // TODO: OPTIM: We currently copy the buffers because the JavaScript representation of
-        //   arrays does not match the Rust implementation. Try to reduce copies where possible by
-        //   reusing it, i.e. check, if `target_len` >= `data_capacity` and constructing it from raw
-        //   parts.
+        if data_len == target_len {
```

Comment:
To me this is the only case where it's ok to not copy on the Rust side.
Since Arrow JS doesn't always write the null values at the end of the buffer, I don't think we can map the array to Rust as we would map some random JS memory.
Even if `target_len <= data_capacity` we would be mapping uninitialized memory and not null values.

Reply:
As we are not freeing the memory in Rust, we could use a slice from the buffer, i.e. something like `slice_from_raw_parts(data_ptr, target_len)` if `target_len <= data_capacity`.

Reply:
Ok going with `target_len <= data_capacity` https://github.com/hashintel/hash/pull/563/commits/fed9ed9b47b9795e2ce25dd1800a323e5b8aa541

I'm not sure what you mean in your second comment.

Reply:
We use a buffer with `target_len` elements, not `data_capacity` len, as you did 🙂